src/automatic_differentiation.jl

```diff
@@ -20,6 +20,14 @@ end
 ####################
 
 # Scalar output -> Scalar value
+"""
+    function _extract_value(v::ForwardDiff.Dual)
```

Comment:
If this (and `_insert_gradient`) are documented, perhaps remove the underscore?

Reply:
They are documented, but I didn't include them in the docs.
(As I documented them before using your @implement_gradient idea). 
Should I then remove underscore, include in docs, and export them, or just remove underscore?